project-koku/koku
# 2003 Recover from Producer.close issue

#2003 

Testing (with docker-compose):
1. Get ingress service working (This is a good start: [https://github.com/project-koku/koku/blob/master/docs/source/testing.rst#testing-with-ingress](https://github.com/project-koku/koku/blob/master/docs/source/testing.rst#testing-with-ingress)).
2. Create a payload to ingest. I used Nise to generate files with an ocp_cluster_id of `my-cluster-id`. Then, I used the new script like this:
```
python scripts/ocp_packager.py 'testing/pvc_dir/insights_local/my-cluster-id/20200401-20200501/266974d9-4d59-4bc2-9f26-e603dc214004_openshift_usage_report.0.csv,testing/pvc_dir/insights_local/my-cluster-id/20200401-20200501/266974d9-4d59-4bc2-9f26-e603dc214004_openshift_usage_report.1.csv' my-cluster-id
```
That command created `payload.tar.gz` in my `koku` directory
3. Create a source that will correspond to the payload.
```
POST http://localhost:8000/api/cost-management/v1/sources/
{
    "name": "ProviderOCP",
    "source_type": "OCP",
    "authentication": {
        "resource_name": "my-cluster-id"
    }
}
```
4. Check that ingress is working (change the `-F "upload=@` to match wherever your payload is):
```
curl -vvvv -H "x-rh-identity: eyJpZGVudGl0eSI6IHsiYWNjb3VudF9udW1iZXIiOiAiMDAwMDAwMSIsICJpbnRlcm5hbCI6IHsib3JnX2lkIjogIjAwMDAwMSJ
9fX0=" -F "upload=@/Users/mskarbek/Documents/code/koku/payload.tar.gz;type=application/vnd.redhat.hccm.tar+tgz" -H "x-rh-insights-request-id: 52df9f748eabcfea" localhost:8080/api/ingress/v1/upload
```
You'll see logs in the koku-listener if things are working correctly.

5. Once you know ingress is working, pepper some sleeps into the `kafka_msg_handler`:
These can be added together or separately. First diff:
```diff
diff --git a/koku/masu/external/kafka_msg_handler.py b/koku/masu/external/kafka_msg_handler.py
index 93844df1..d426e133 100644
--- a/koku/masu/external/kafka_msg_handler.py
+++ b/koku/masu/external/kafka_msg_handler.py
@@ -177,6 +177,8 @@ async def send_confirmation(request_id, status):  # pragma: no cover
         None
 
     """
+    LOG.warning("SLEEPING...")
+    await asyncio.sleep(10)
     producer = AIOKafkaProducer(loop=EVENT_LOOP, bootstrap_servers=Config.INSIGHTS_KAFKA_ADDRESS)
     try:
         await producer.start()
```
second diff:
```diff
diff --git a/koku/masu/external/kafka_msg_handler.py b/koku/masu/external/kafka_msg_handler.py
index 93844df1..6547d4f2 100644
--- a/koku/masu/external/kafka_msg_handler.py
+++ b/koku/masu/external/kafka_msg_handler.py
@@ -329,6 +329,8 @@ async def process_messages():  # pragma: no cover
             value = json.loads(msg.value.decode("utf-8"))
             count = 0
             while True:
+                LOG.warning("SLEEPING...")
+                await asyncio.sleep(10)
                 try:
                     await send_confirmation(value["request_id"], status)
                     break
```
6. Restart koku-listener
7. Post the payload again. _**During the sleep, turn off kafka**_. Watch the koku-listener logs.
```
koku_listener     | [2020-04-03 18:23:26,969] ERROR None Unable to connect to kafka server.  Closing producer. ConnectionError: Unable to bootstrap from [('kafka', 29092, <AddressFamily.AF_UNSPEC: 0>)]
koku_listener     | Traceback (most recent call last):
koku_listener     |   File "/koku/koku/masu/external/kafka_msg_handler.py", line 182, in send_confirmation
koku_listener     |     await producer.start()
koku_listener     |   File "/opt/app-root/lib/python3.6/site-packages/aiokafka/producer/producer.py", line 286, in start
koku_listener     |     yield from self.client.bootstrap()
koku_listener     |   File "/opt/app-root/lib/python3.6/site-packages/aiokafka/client.py", line 228, in bootstrap
koku_listener     |     'Unable to bootstrap from {}'.format(self.hosts))
koku_listener     | kafka.errors.ConnectionError: ConnectionError: Unable to bootstrap from [('kafka', 29092, <AddressFamily.AF_UNSPEC: 0>)]
koku_listener     | [2020-04-03 18:23:26,978] ERROR None Resending message confirmation due to error: Unable to connect to kafka server.  Closing producer.
koku_listener     | [2020-04-03 18:23:26,978] INFO None Sleeping for 1.02 seconds.
koku_listener     | [2020-04-03 18:23:28,006] WARNING None SLEEPING... <<<<<<<<<<<<< this is from the diff
```
8. Restart kafka. Watch the logs recover and see the message sent:
```
koku_listener     | [2020-04-03 18:24:05,317] INFO None Validating message: b'{"request_id": "52df9f748eabcfea", "validation": "success"}'
koku_listener     | [2020-04-03 18:24:05,329] INFO None Validating message complete.
```
